src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -329,15 +338,25 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                     return;
                 }
 
-                // For other mutable interfaces like ICollection<>, IDictionary<,> and ISet<>, we prefer copying values and setting them
-                // on a new instance of the interface over populating the existing instance implementing the interface.
-                // This has already been done, so there's not need to check again.
-                if (TypeIsADictionaryInterface(type) && !bindingPoint.IsReadOnly)
+                // Handle IDictionary<TKey, TValue> and IReadOnlyDictionary<TKey, TValue> cases.
+                // if we have IDictionary<TKey, TValue>:
+                //      - if bindingPoint.Value is not null, we'll use this instance to populate the configuration items to it.
+                //      - if bindingPoint.Value is null and bindingPoint.IsReadOnly is false (i.e. there is a property setter). we'll create
+                //        a Dictionary<TKey, TValue> and populate the configuration items to it.
+                //      - otherwise nothing will happen.
+                // if we have IReadOnlyDictionary<TKey, TValue>:
+                //      - if bindingPoint.Value is not null and bindingPoint.IsReadOnly is false, we'll create a new Dictionary<TKey, TValue> and
```

Comment:
1. I think this would read easier if it didn't try to be pseudocode. `if bindingPoint.Value is not null and bindingPoint.IsReadOnly is false` => `if the property value is not null and the property has a setter`.
2. What about when the value is `null` and there is a setter? This comment makes it sound like nothing will happen.